File: backend/app/gst_engine.py
"""GST computation engine — Python port of src/lib/gstEngine.ts.
Kept identical so frontend previews and backend invoices always match.
"""
from dataclasses import dataclass, field

STATE_CODES = {
    "01": "Jammu and Kashmir", "02": "Himachal Pradesh", "03": "Punjab", "04": "Chandigarh",
    "05": "Uttarakhand", "06": "Haryana", "07": "Delhi", "08": "Rajasthan", "09": "Uttar Pradesh",
    "10": "Bihar", "19": "West Bengal", "24": "Gujarat", "27": "Maharashtra", "29": "Karnataka",
    "32": "Kerala", "33": "Tamil Nadu", "36": "Telangana", "37": "Andhra Pradesh",
}
NAME_TO_CODE = {v.lower(): k for k, v in STATE_CODES.items()}


def code_for_state(name: str | None) -> str | None:
    return NAME_TO_CODE.get((name or "").strip().lower())


def _round2(n: float) -> float:
    return round(n + 1e-9, 2)


@dataclass
class Computation:
    taxable_value: float = 0
    cgst: float = 0
    sgst: float = 0
    igst: float = 0
    total_tax: float = 0
    round_off: float = 0
    grand_total: float = 0
    is_inter_state: bool = False
    place_of_supply: str = ""
    amount_in_words: str = ""
    lines: list = field(default_factory=list)


def resolve_supply(seller_state: str, seller_gstin: str, buyer_gstin: str | None, buyer_state: str | None):
    seller_code = (seller_gstin[:2] if seller_gstin else None) or code_for_state(seller_state) or "09"
    buyer_code = None
    if buyer_gstin and len(buyer_gstin) >= 2 and buyer_gstin[:2] in STATE_CODES:
        buyer_code = buyer_gstin[:2]
    elif buyer_state:
        buyer_code = code_for_state(buyer_state)
    is_inter = bool(buyer_code) and buyer_code != seller_code
    pos_code = buyer_code or seller_code
    pos_name = STATE_CODES.get(pos_code, seller_state)
    return is_inter, f"{pos_code} - {pos_name}"


def compute(items: list[dict], is_inter_state: bool, place_of_supply: str) -> Computation:
    c = Computation(is_inter_state=is_inter_state, place_of_supply=place_of_supply)
    for it in items:
        taxable = _round2(it["qty"] * it["rate"])
        tax = _round2(taxable * it["gstRate"] / 100)
        cgst = 0 if is_inter_state else _round2(tax / 2)
        sgst = 0 if is_inter_state else _round2(tax - cgst)
        igst = tax if is_inter_state else 0
        c.taxable_value += taxable
        c.cgst += cgst
        c.sgst += sgst
        c.igst += igst
        c.lines.append({"taxable": taxable, "tax": tax})
    c.taxable_value = _round2(c.taxable_value)
    c.cgst, c.sgst, c.igst = _round2(c.cgst), _round2(c.sgst), _round2(c.igst)
    c.total_tax = _round2(c.cgst + c.sgst + c.igst)
    raw = c.taxable_value + c.total_tax
    c.grand_total = round(raw)
    c.round_off = _round2(c.grand_total - raw)
    c.amount_in_words = number_to_words(c.grand_total)
    return c


def next_invoice_number(prefix: str, existing: list[str]) -> str:
    from datetime import datetime
    now = datetime.now()
    start = now.year if now.month >= 4 else now.year - 1
    fy = f"{start}-{str((start + 1) % 100).zfill(2)}"
    mx = 0
    for n in existing:
        digits = "".join(ch for ch in n.split("/")[-1] if ch.isdigit())
        if digits:
            mx = max(mx, int(digits))
    return f"{prefix.rstrip('/')}/{fy}/{str(mx + 1).zfill(4)}"


_ONES = ["", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten",
         "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen"]
_TENS = ["", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"]


def _two(n: int) -> str:
    if n < 20:
        return _ONES[n]
    return _TENS[n // 10] + ((" " + _ONES[n % 10]) if n % 10 else "")


def number_to_words(num: float) -> str:
    num = round(num)
    if num == 0:
        return "Zero Rupees Only"
    words = ""
    for div, label in ((10000000, "Crore"), (100000, "Lakh"), (1000, "Thousand"), (100, "Hundred")):
        q, num = divmod(num, div)
        if q:
            words += (_two(q) if div != 100 else _ONES[q]) + f" {label} "
    if num:
        words += ("and " if words else "") + _two(num) + " "
    return words.strip() + " Rupees Only"

File: android/lib/gstEngine.ts
// AK-LOGIC AI GST — GST Engine for Android Client
// 100% compliant with Indian Goods & Services Tax rules

export interface InvoiceItem {
  id?: string;
  description: string;
  hsn?: string;
  qty: number;
  rate: number;
  gstRate: number;
  inventoryItemId?: string;
}

export interface LineComputation {
  item: InvoiceItem;
  taxable: number;
  taxAmount: number;
  cgst: number;
  sgst: number;
  igst: number;
  total: number;
}

export interface InvoiceComputation {
  lines: LineComputation[];
  taxableValue: number;
  totalCgst: number;
  totalSgst: number;
  totalIgst: number;
  totalTax: number;
  roundOff: number;
  grandTotal: number;
  isInterState: boolean;
  placeOfSupply: string;
  amountInWords: string;
  hsnSummary: { hsn: string; taxable: number; gstRate: number; cgst: number; sgst: number; igst: number }[];
}

export const STATE_CODES: Record<string, string> = {
  '01': 'Jammu and Kashmir', '02': 'Himachal Pradesh', '03': 'Punjab', '04': 'Chandigarh',
  '05': 'Uttarakhand', '06': 'Haryana', '07': 'Delhi', '08': 'Rajasthan',
  '09': 'Uttar Pradesh', '10': 'Bihar', '11': 'Sikkim', '12': 'Arunachal Pradesh',
  '13': 'Nagaland', '14': 'Manipur', '15': 'Mizoram', '16': 'Tripura',
  '17': 'Meghalaya', '18': 'Assam', '19': 'West Bengal', '20': 'Jharkhand',
  '21': 'Odisha', '22': 'Chhattisgarh', '23': 'Madhya Pradesh', '24': 'Gujarat',
  '26': 'Dadra and Nagar Haveli and Daman and Diu', '27': 'Maharashtra', '28': 'Andhra Pradesh',
  '29': 'Karnataka', '30': 'Goa', '31': 'Lakshadweep', '32': 'Kerala',
  '33': 'Tamil Nadu', '34': 'Puducherry', '35': 'Andaman and Nicobar Islands', '36': 'Telangana',
  '37': 'Andhra Pradesh (New)', '38': 'Ladakh',
};

export const INDIAN_STATES: string[] = [
  'Andaman and Nicobar Islands', 'Andhra Pradesh', 'Arunachal Pradesh', 'Assam', 'Bihar',
  'Chandigarh', 'Chhattisgarh', 'Dadra and Nagar Haveli and Daman and Diu', 'Delhi', 'Goa',
  'Gujarat', 'Haryana', 'Himachal Pradesh', 'Jammu and Kashmir', 'Jharkhand', 'Karnataka',
  'Kerala', 'Ladakh', 'Lakshadweep', 'Madhya Pradesh', 'Maharashtra', 'Manipur', 'Meghalaya',
  'Mizoram', 'Nagaland', 'Odisha', 'Puducherry', 'Punjab', 'Rajasthan', 'Sikkim', 'Tamil Nadu',
  'Telangana', 'Tripura', 'Uttar Pradesh', 'Uttarakhand', 'West Bengal'
];

export function codeForStateName(stateName?: string | null): string | null {
  if (!stateName) return null;
  const target = stateName.trim().toLowerCase();
  for (const [code, name] of Object.entries(STATE_CODES)) {
    if (name.toLowerCase() === target) return code;
  }
  return null;
}

export function stateNameFromGstin(gstin?: string | null): string | null {
  if (!gstin || gstin.length < 2) return null;
  const code = gstin.slice(0, 2);
  return STATE_CODES[code] || null;
}

function round2(n: number) {
  return Math.round(n * 100) / 100;
}

export function numberToWords(num: number): string {
  const a: string[] = ['', 'One ', 'Two ', 'Three ', 'Four ', 'Five ', 'Six ', 'Seven ', 'Eight ', 'Nine ', 'Ten ', 'Eleven ', 'Twelve ', 'Thirteen ', 'Fourteen ', 'Fifteen ', 'Sixteen ', 'Seventeen ', 'Eighteen ', 'Nineteen '];
  const b: string[] = ['', '', 'Twenty', 'Thirty', 'Forty', 'Fifty', 'Sixty', 'Seventy', 'Eighty', 'Ninety'];

  const n = ('000000000' + num).substr(-9).match(/^(\d{2})(\d{2})(\d{2})(\d{1})(\d{2})$/);
  if (!n) return '';
  let str = '';
  str += Number(n[1]) !== 0 ? (a[Number(n[1])] || b[Number(n[1][0])] + ' ' + a[Number(n[1][1])]) + 'Crore ' : '';
  str += Number(n[2]) !== 0 ? (a[Number(n[2])] || b[Number(n[2][0])] + ' ' + a[Number(n[2][1])]) + 'Lakh ' : '';
  str += Number(n[3]) !== 0 ? (a[Number(n[3])] || b[Number(n[3][0])] + ' ' + a[Number(n[3][1])]) + 'Thousand ' : '';
  str += Number(n[4]) !== 0 ? (a[Number(n[4])] || b[Number(n[4][0])] + ' ' + a[Number(n[4][1])]) + 'Hundred ' : '';
  str += Number(n[5]) !== 0 ? ((str !== '') ? 'and ' : '') + (a[Number(n[5])] || b[Number(n[5][0])] + ' ' + a[Number(n[5][1])]) : '';
  return 'INR ' + str.trim() + ' Only';
}

export function resolveSupply(opts: {
  sellerState: string;
  sellerGstin: string;
  buyerGstin?: string;
  buyerState?: string;
}): { isInterState: boolean; placeOfSupply: string } {
  const sellerCode = opts.sellerGstin?.slice(0, 2) || codeForStateName(opts.sellerState) || '27';
  const sellerName = STATE_CODES[sellerCode] || opts.sellerState || 'Maharashtra';

  let buyerName: string | null = stateNameFromGstin(opts.buyerGstin);
  if (!buyerName && opts.buyerState) buyerName = opts.buyerState;
  const buyerCode = buyerName ? (codeForStateName(buyerName) || opts.buyerGstin?.slice(0, 2) || null) : null;

  const isInterState = !!buyerCode && buyerCode !== sellerCode;
  const posCode = buyerCode || sellerCode;
  const posName = STATE_CODES[posCode] || sellerName;

  return {
    isInterState,
    placeOfSupply: `${posCode} - ${posName}`,
  };
}

export function computeInvoice(
  items: InvoiceItem[],
  ctx: { isInterState: boolean; placeOfSupply: string }
): InvoiceComputation {
  const lines: LineComputation[] = items.map((item) => {
    const taxable = round2((item.qty || 0) * (item.rate || 0));
    const taxAmount = round2((taxable * (item.gstRate || 0)) / 100);
    const cgst = ctx.isInterState ? 0 : round2(taxAmount / 2);
    const sgst = ctx.isInterState ? 0 : round2(taxAmount - cgst);
    const igst = ctx.isInterState ? taxAmount : 0;
    return { item, taxable, taxAmount, cgst, sgst, igst, total: round2(taxable + taxAmount) };
  });

  const taxableValue = round2(lines.reduce((s, l) => s + l.taxable, 0));
  const totalCgst = round2(lines.reduce((s, l) => s + l.cgst, 0));
  const totalSgst = round2(lines.reduce((s, l) => s + l.sgst, 0));
  const totalIgst = round2(lines.reduce((s, l) => s + l.igst, 0));
  const totalTax = round2(totalCgst + totalSgst + totalIgst);

  const rawTotal = taxableValue + totalTax;
  const grandTotal = Math.round(rawTotal);
  const roundOff = round2(grandTotal - rawTotal);

  const hsnMap = new Map<string, { hsn: string; taxable: number; gstRate: number; cgst: number; sgst: number; igst: number }>();
  for (const l of lines) {
    const key = `${l.item.hsn || '-'}_${l.item.gstRate}`;
    const cur = hsnMap.get(key) || { hsn: l.item.hsn || '-', taxable: 0, gstRate: l.item.gstRate, cgst: 0, sgst: 0, igst: 0 };
    cur.taxable = round2(cur.taxable + l.taxable);
    cur.cgst = round2(cur.cgst + l.cgst);
    cur.sgst = round2(cur.sgst + l.sgst);
    cur.igst = round2(cur.igst + l.igst);
    hsnMap.set(key, cur);
  }

  return {
    lines,
    taxableValue,
    totalCgst,
    totalSgst,
    totalIgst,
    totalTax,
    roundOff,
    grandTotal,
    isInterState: ctx.isInterState,
    placeOfSupply: ctx.placeOfSupply,
    amountInWords: numberToWords(grandTotal),
    hsnSummary: Array.from(hsnMap.values()),
  };
}

export function nextInvoiceNumber(prefix: string, existingNumbers: string[]): string {
  const now = new Date();
  const month = now.getMonth();
  const startYear = month >= 3 ? now.getFullYear() : now.getFullYear() - 1;
  const fy = `${startYear}-${String((startYear + 1) % 100).padStart(2, '0')}`;
  const base = (prefix || 'INV').replace(/\/+$/, '');
  let max = 0;
  for (const n of existingNumbers) {
    const m = n.match(/(\d+)\s*$/);
    if (m) max = Math.max(max, parseInt(m[1], 10));
  }
  const seq = String(max + 1).padStart(4, '0');
  return `${base}/${fy}/${seq}`;
}
